server/apps/expand.go

```diff
@@ -1,182 +1,209 @@
+	if expand == nil {
+		clone.expandedContext = expandedContext{}
+		return &clone, nil
+	}
+
+	if expand.ActingUser != "" && e.ActingUserID != "" && e.ActingUser == nil {
+		actingUser, err := e.s.Mattermost.User.Get(e.ActingUserID)
 		if err != nil {
-			return nil, err
+			return nil, errors.Wrapf(err, "failed to expand acting user %s", e.ActingUserID)
 		}
+		e.ActingUser = actingUser
 	}
 
-	expanded = e.produce(expand)
-	return expanded, nil
-}
+	if expand.App != "" && e.AppID != "" && e.App == nil {
+		app, err := e.s.Store.GetApp(e.AppID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to expand app %s", e.AppID)
+		}
+		e.App = app
+	}
 
-func (e *expander) collectConfig(expand *Expand) error {
-	if e.Config != nil || !expand.Config {
-		return nil
+	if expand.Channel != "" && e.ChannelID != "" && e.Channel == nil {
+		ch, err := e.s.Mattermost.Channel.Get(e.ChannelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to expand channel %s", e.ChannelID)
+		}
+		e.Channel = ch
 	}
-	e.Config = e.configurator.GetMattermostConfig()
-	return nil
-}
 
-func (e *expander) collectChannel(channelID string) func(*Expand) error {
-	return func(expand *Expand) error {
-		if channelID == "" || !isValidExpandLevel(expand.Channel) {
-			return nil
+	// Config is cached pre-sanitized
+	if expand.Config && e.Config == nil {
+		mmconf := e.s.Configurator.GetMattermostConfig()
+		e.Config = &MattermostConfig{}
+		if mmconf.ServiceSettings.SiteURL != nil {
+			e.Config.SiteURL = *mmconf.ServiceSettings.SiteURL
 		}
+	}
+
+	// TODO expand Mentioned
 
-		mmchannel, err := e.mm.Channel.Get(channelID)
+	if expand.Post != "" && e.PostID != "" && e.Post == nil {
+		post, err := e.s.Mattermost.Post.GetPost(e.PostID)
 		if err != nil {
-			return err
+			return nil, errors.Wrapf(err, "failed to expand post %s", e.PostID)
 		}
+		e.Post = post
+	}
 
-		e.Channel = mmchannel
-		return nil
+	if expand.RootPost != "" && e.RootPostID != "" && e.RootPost == nil {
+		post, err := e.s.Mattermost.Post.GetPost(e.RootPostID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to expand root post %s", e.RootPostID)
+		}
+		e.RootPost = post
 	}
-}
 
-func (e *expander) collectUser(userID string, userref **model.User) func(*Expand) error {
-	return func(expand *Expand) error {
-		if *userref != nil || userID == "" || !isValidExpandLevel(expand.User) {
-			return nil
+	if expand.Team != "" && e.TeamID != "" && e.Team == nil {
+		team, err := e.s.Mattermost.Team.Get(e.TeamID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to expand team %s", e.TeamID)
 		}
+		e.Team = team
+	}
 
-		mmuser, err := e.mm.User.Get(userID)
+	if expand.User != "" && e.UserID != "" && e.User == nil {
+		user, err := e.s.Mattermost.User.Get(e.UserID)
 		if err != nil {
-			return err
+			return nil, errors.Wrapf(err, "failed to expand user %s", e.UserID)
 		}
-		mmuser.SanitizeProfile(nil)
+		e.User = user
+	}
+
+	clone.expandedContext = expandedContext{
```

Comment:
Interesting implementation. Send in the full model to each strip method return cloned model, or reduced cloned model if `store.ExpandSummary`.  I like this. 👍